src/crowdin_client.cpp

```diff
@@ -50,64 +57,71 @@
     using std::smatch;
 #endif
 
+
+
 // ----------------------------------------------------------------
 // Implementation
 // ----------------------------------------------------------------
 
 namespace
 {
 
-#define OAUTH_CLIENT_ID     "poedit-y1aBMmDbm3s164dtJ4Ur150e2"
-#define OAUTH_AUTHORIZE_URL "/oauth2/authorize?response_type=token&client_id=" OAUTH_CLIENT_ID
+#define OAUTH_SCOPE         "project"
+#define OAUTH_CLIENT_ID     "k0uFz5HYQh0VzWgZmOpA"
+//      any arbitrary unique unguessable string (e.g. UUID in hex)
+#define OAUTH_STATE         "948cf13ffffb47119d6cfa2b68898f67"
+//      The value of below macro should be set exactly as is (without quotes)
+//      to "Authorization Callback URL" of Crowdin application
+//      https://support.crowdin.com/enterprise/creating-oauth-app/
 #define OAUTH_URI_PREFIX    "poedit://auth/crowdin/"
-
-// Recursive extract files from /api/project/*/info response
-void ExtractFilesFromInfo(std::vector<std::wstring>& out, const json& r, const std::wstring& prefix)
-{
-    for (auto i : r["files"])
-    {
-        std::wstring name = prefix + str::to_wstring(i["name"]);
-        std::string node_type = i["node_type"];
-        if (node_type == "file")
-        {
-            out.push_back(name);
-        }
-        else if (node_type == "directory" || node_type == "branch")
-        {
-            ExtractFilesFromInfo(out, i, name + L"/");
-        }
-    }
-}
+#define OAUTH_AUTHORIZE_URL "/oauth/authorize" \
+    "?" "response_type" "=" "token" \
+    "&" "scope"         "=" OAUTH_SCOPE \
+    "&" "client_id"     "=" OAUTH_CLIENT_ID \
+    "&" "state"         "=" OAUTH_STATE \
+    "&" "redirect_uri"  "=" OAUTH_URI_PREFIX \
+    "&" "utm_source=poedit.net&utm_medium=referral&utm_campaign=poedit" // source tracking
 
 } // anonymous namespace
 
-
 std::string CrowdinClient::WrapLink(const std::string& page)
 {
-    std::string url("https://poedit.net/crowdin");
+    std::string url("https://accounts.crowdin.com");
     if (!page.empty() && page != "/")
-        url += "?u=" + http_client::url_encode(page);
+        url += page;
     return url;
 }
 
 
 class CrowdinClient::crowdin_http_client : public http_client
 {
 public:
-    crowdin_http_client(CrowdinClient& owner)
-        : http_client("https://api.crowdin.com"), m_owner(owner)
+    crowdin_http_client(CrowdinClient& owner, const std::string& url_prefix)
+        : http_client(url_prefix), m_owner(owner)
     {}
 
 protected:
     std::string parse_json_error(const json& response) const override
     {
-        std::string msg = response["error"]["message"];
-
-        // Translate commonly encountered messages:
-        if (msg == "Translations download is forbidden by project owner")
-            msg = _("Downloading translations is disabled in this project.").utf8_str();
```

Comment:
Does the removal of this mean that this situation can no longer happen in V2 of the API?

Reply:
It might, but so far I realized it might happen by at least more than one different reasons so displaying message from server (i.e. in response of API request) as is usually more informative here. As well as there might be some other reasons existing now or added in the future we are just not aware about so we might just to miss to catch all of them in any case if to try to generalize

Reply:
Displaying server's error is of course best. But I think you are misinterpreting this code: it doesn't hide server error, it replaces *one exact specific text*, with another. The reason was that

(a) in V1 api, that the error given by the server was cryptic (people didn't understand that "forbidden" is something they just need to enable in the settings and interpreted it as being forbidden from doing it by Crowdin or by Poedit's free version etc.); and

(b) this code ensure *translation* of by far the most common error message. I don't think Crowdin localizes their errors per `Accept-Language` but maybe they do now…

In any case, I can easily check if this error is still produced by V2 or not.